[REVIEW] Add documentation on libcudf doxygen guidelines [skip ci] (#5100)
rapidsai/cudf: cpp/docs/DOCUMENTATION.md

# libcudf++ C++ Documentation Guide

This document is a guide for documenting the libcudf++ source code that is rendered by the
doxygen tool and published to our external RAPIDS [web site](https://docs.rapids.ai/api/libcudf/stable/index.html).

## Copyright License

This is included here but may also be mentioned in a coding guideline document as well.
The following is the license header comment that should appear at the beginning of every C++ source file.

```c++
/*
 * Copyright (c) 2020, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

> The comment should start with `\*` and not `\**` so it is not processed by doxygen.

Reviewer: ```suggestion
The comment should start with `/*` and not `/**` so it is not processed by doxygen.
```

Each line in the comment block between the `/**` and `*/` lines should start with a space followed by an asterisk. Any text on these lines, including tag declarations, should start after a single space after the asterisk.

## Tag/Command names

Use `@` to prefix doxygen commands (e.g. `@brief`, `@code`, etc.)

## Markdown

The doxygen tool supports a limited set of markdown format in the comment block including links, tables, lists, etc.

In some cases a trade-off may be required for readability in the source text file versus the readability in the doxygen formatted web pages.

For example, there are some limitations on readability with '%' character and pipe character '|' within a markdown table.

Also, try to avoid using direct hTML tags. Although doxygen supports markdown and markdown supports HTML tags, the HTML support for doxygen's markdown is also limited.

## The Example

The following example will cover most of the doxygen block comment and tag styles
for documenting C++ code in libcudf.

```c++
/**
 * @file source_file.cpp
 * @brief Description of source file contents
 *
 * Longer description of the source file contents.
 */

> 

Reviewer 2: Maybe add a group definition example here?

Author: The `@ingroup` example is included in the comment below.

/**
 * @brief One sentence description of the class.
 *
 * @ingroup optional_group_name_for_modules_page
 *
 * Longer, more detailed desriptio of the class.
 *
 * @tparam T Short description of any template parameters
 */
template <typename T>
class example_class {

  void get_my_int();            ///< Members can be documented like this
  void set_my_int( int value ); ///< Try to use descriptive member names

  /**
   * @brief Short, one sentence description of the member function.
   *
   * A more detailed description of what this function does and what
   * its logic does.
   *
   * @code
   * example_class<int> inst;
   * inst.set_my_int(5);
   * int output = inst.complicated_function(1,dptr,fptr);
   * @endcode
   *
   * @param[in]     first  This parameter is an input parameter to the function
   * @param[in,out] second This parameter is used both as an input and output
   * @param[out]    third  This parameter is an output of the function
   *
   * @return The result of the complex function
   */
  T complicated_function(int first, double* second, float* third)
  {
      // Do not use doxygen-style block comments
      // for code logic documentation.
  }

 private:

>   int my_int;                ///< An example private member variable

Author: Perhaps useful information to someone using the class?

  return functor(input_argument);
}

/**
 * @brief Short, one sentence description.
 *
 * @ingroup optional_one_or_more predefined_groups
 *
 * Optional, longer description.
 */
enum class example_enum {
  first_enum,   ///< Description of the first enum
  second_enum,  ///< Description of the second enum
  third_enum    ///< Description of the third enum
};
```

## Descriptions

The comment description should clearly detail how the output(s) are created from any inputs.
Include any performance and any boundary considerations.
Also include any limits on parameter values and if any default values are declared.
Don't forget to specify how nulls are handled or produced.
Also, try to include a short [example](#inline-examples) if possible.

### @brief

The `@brief` text should be a short, one sentence description.
Doxygen does not provide much space to show this text in the output pages.
Always follow the `@brief` line with a blank comment line.

The longer description is the rest of the comment text that is not tagged with any doxygen command.

```c++
/**
 * @brief Short description.
 *
 * Long description.
 *
```

### @copydoc

> Documentation for declarations in headers is expected to be clear and complete.

Reviewer 2: ```suggestion
Documentation for declarations in headers should be clear and complete.
```

> You can use the `@copydoc` tag to avoid duplicating the comment block for a function definition for example.

Reviewer 2: ```suggestion
You can use the `@copydoc` tag to avoid duplicating the comment block for a function definition.
```

```c++
  /**
   * @copydoc complicated_function(int,double*,float*)
   *
   * Any extra documentation.
   */
```

> Also, the `@copydoc` is useful when documenting a `detail` function that differs only by the `cudaStream_t` parameter.

Reviewer 2: ```suggestion
Also, `@copydoc` is useful when documenting a `detail` function that differs only by the `cudaStream_t` parameter.
```

| [@param](#param) | Description for each function parameter |
| [@return](#return) | Short description of object or value returned |

#### @throw

Add an [@throw](http://www.doxygen.nl/manual/commands.html#cmdthrow) comment line in the comment block for each exception that the function may throw.
You only need to include exceptions thrown by the function itself.
If the function calls another function that may throw an exception, you do not need to document those exceptions here.

Include the name of the exception without tick marks so doxygen can add reference links correctly.

```c++
 *
 * @throw cudf::logic_error if `input_argument` is negative or zero
 *
```

Using `@throws` is also acceptable but vs-code and other tools only do syntax highlighting on `@throw`.

#### @tparam

Add a [@tparam](http://www.doxygen.nl/manual/commands.html#cmdtparam) comment line for each template parameter declared by this function.
The name of the parameter in the comment must match exactly to the template parameter name.

```c++
 *
 * @tparam functor_type The type of the functor
 * @tparam input_type The datatype of the input argument
 *
```

> The definition should detail the requirements of parameter.

Reviewer 2: ```suggestion
The definition should detail the requirements of the parameter.
```

```c++
 *
 * @param[in]     first  This parameter is an input parameter to the function
 * @param[in,out] second This parameter is used both as an input and output
 * @param[out]    third  This parameter is an output of the function
 *
```

Also recommend vertically aligning the 3 columns of text if possible to make it easier to read in the source code itself.

#### @return

Add a single [@return](http://www.doxygen.nl/manual/commands.html#cmdreturn) comment line at the end of the comment block if the function returns an object or value.
Include a brief description of what is returned.

```c++
/**
 * ...
 *
 * @return A new column of type INT32 and no nulls
 */
```

Do not include the type of the object returned with the `@return` comment.

### Inline Examples

It is usually helpful to include a source code example inside your comment block when documenting a function or other declaration.
Use the [@code](http://www.doxygen.nl/manual/commands.html#cmdcode) and [@endcode](http://www.doxygen.nl/manual/commands.html#cmdendcode) pair to include inline examples.

Doxygen supports syntax highlighting for C++ and several other programming languages (e.g. Python, Java).

> By default, the `@code` tag will use syntax highlighting based on the source code where it was found.

Reviewer 2: ```suggestion
By default, the `@code` tag uses syntax highlighting based on the source code in which it is found.
```

```c++
 *
 * @code
 * auto result = cudf::make_column( );
 * @endcode
 *
```

You can specify a different language by indicating the file extension in the tag:

```c++
 *
 * @code{.py}
 * import cudf
 * s = cudf.Series([1,2,3])
 * @endcode
 *
```

If you wish to use pseudo-code in your example, use the following:

```c++
 *
 * Sometimes pseudo-code is clearer.
 * @code{.pseudo}
 * s = int column of [ 1, 2, null, 4 ]
 * r = fill( s, [1, 2], 0 )
 * r is now [ 1, 0, 0, 4 ]
 * @endcode
 *
```

When writing example snippets, using fully qualified class names allows doxygen to add reference links to the example.

```c++
 *
 * @code
 * auto result1 = make_column( ); // reference link will not be created
 * auto result2 = cudf::make_column( ); // reference link will be created
 * @endcode
 *
```

Although using 3 backtick marks `` ``` `` for example blocks will work too, they do not standout as well in vs-code and other source editors.

> Do not use the `@example` tag in the comments for a declaration.
> Doxygen will interpret the entire source file as example source code when it encounters this tag.
> The source file is then published under a separate the _Examples_ page in the output.

Reviewer 2: ```suggestion
Do not use the `@example` tag in the comments for a declaration, or Doxygen will interpret the entire source file as example source code. The source file is then published under a separate _Examples_ page in the output.
```

## Namespaces

Doxygen output includes a _Namespaces_ page that shows all the namespaces declared with comment blocks in the processed files.

> Here is an example of doxygen description comment for a namespace declaration.

Reviewer 2: ```suggestion
Here is an example doxygen description comment for a namespace declaration.
```

```c++

> //! cuDF interfaces

Reviewer 2: Shouldn't any namespace we document include more detail and therefore deserve a `/**` doxygen block?

namespace cudf {
```

> A description comment should be included for only once for each unique namespace declaration.

Reviewer 2: ```suggestion
A description comment should be included only once for each unique namespace declaration.
```

> Otherwise, if more than one description is found, doxygen will aggregate the descriptions in an arbitrary order in the output pages.

Reviewer 2: ```suggestion
Otherwise, if more than one description is found, doxygen aggregates the descriptions in an arbitrary order in the output pages.
```

> If you are introducing a new namespace, provide a short description block for only one declaration and not for every occurrence.

Reviewer 2: ```suggestion
If you introduce a new namespace, provide a short description block for only one declaration and not for every occurrence.
```

## Groups/Modules

Grouping declarations into modules helps users to find APIs in the doxygen pages.

> Generally, common functions are already grouped logically into header files but not easily reflected in the doxygen output.

Reviewer 2: ```suggestion
Generally, common functions are already grouped logically into header files but doxygen does not automatically group them in its output.
```

> Doxygen output includes a _Modules_ page that organizes items into groups using the [Grouping doxygen commands](http://www.doxygen.nl/manual/grouping.html)

Reviewer 2: ```suggestion
Doxygen output includes a _Modules_ page that organizes items into groups specified using the [Grouping doxygen commands](http://www.doxygen.nl/manual/grouping.html)
```

These commands can group common functions across header files, source files, and even namespaces.
Groups can also be nested by defining new groups within existing groups.

For libcudf, all the group hierarchy is defined in the `doxygen_groups.h` header file.

> This file does not need to be included in any other source file.
> It is processed only by the doxygen tool.
> Use this file to add or update any groups.
> Doxygen will also use the definitions in this file to order to groups in the _Modules_ page.

Reviewer 2: ```suggestion
`doxygen_groups.h` does not need to be included in any other source file, because the definitions in this file are used only by the doxygen tool to generate groups in the _Modules_ page. Modify this file only to add or update groups. The existing groups have been carefully structured and named, so new groups should be added thoughtfully.
```

> When creating a new API, specify the group in which it should belong using the [@ingroup](http://www.doxygen.nl/manual/commands.html#cmdingroup) tag along with the group reference id from the `doxygen_groups.h` file.

Reviewer 2: ```suggestion
When creating a new API, specify its group using the [@ingroup](http://www.doxygen.nl/manual/commands.html#cmdingroup) tag and the group reference id from the `doxygen_groups.h` file.
```

```c++
namespace cudf {

/**
 * @brief ...
 *
 * @ingroup transformation_fill
 *
 * @param ...
 * @return ...
 */
std::unique_ptr<column> fill(table_view const& input,...);

}  // namespace cudf
```

The `@{` is a convenience to include all doxygen documentation from here to the `@}` line as part of the group.

```c++
namespace cudf {
/**
 * @ingroup transformation_fill
 * @{
 */

/**
 * @brief ...
 *
 * @ingroup transformation_fill
 *
 * @param ...
 * @return ...
 */
std::unique_ptr<column> fill(table_view const& input,...);

/** @} */
}  // namespace cudf
```

> This just saves adding `@ingroup` to each comment block that appears in the file.

Reviewer 2: ```suggestion
This saves adding `@ingroup` to each comment block that appears in the file.
```

> Make sure there is a blank line after the group definition comment block so doxygen know it does not apply to whatever follows in the source code.

Reviewer 2: ```suggestion
Make sure there is a blank line after the group definition comment block so doxygen knows it does not apply to whatever follows in the source code.
```

> Note that the `@{` and `@}` does not work especially if it crosses namespaces or includes classes or enums.

Reviewer 2: This sentence is unclear. What exactly doesn't work? You can't have a class or enum definition inside `@{`?

Author: Yes. It looks like it has trouble just with namespace declarations. I reworded this section.

Reviewer 2: Still confusing. "Does not work especially when" means it doesn't work at all, and it especially doesn't work when... :)

Just drop the word especially and it is fixed. :)

## Build Doxygen Output

> The doxygen tool can be installed using the instructions on its [Installation](http://www.doxygen.nl/manual/install.html) page.

Reviewer 2: These instructions are for building from source. I would use conda or apt to install

```suggestion

We recommend installing Doxygen using conda (`conda install doxygen`) or a Linux package manager (`sudo apt install doxygen`). Alternatively you can [build and install from source](http://www.doxygen.nl/manual/install.html).
```

> Building the output is simply running the `doxygen` command while in the `cpp/doxygen` directory containing the `Doxyfile`. The tool will read and process all appropriate source files under the `cpp/include/` directory. The output will be created in the `cpp/doxygen/html/` directory. You can load the local `index.html` file generated there into any web browser to view the result.

Reviewer 2: ```suggestion
To build the libcudf HTML documentatiopn simply run the `doxygen` command from the `cpp/doxygen` directory containing the `Doxyfile`. Doxygen reads and processes all appropriate source files under the `cpp/include/` directory. The output is generated in the `cpp/doxygen/html/` directory. You can load the local `index.html` file generated there into any web browser to view the result.
```

> 

Reviewer 2: ```suggestion

To view docs built on a remote server, you can run a simple HTTP server using Python: `cd html && python -m http.server`. Then open `http://<IP address>:8000` in your local web browser, inserting the IP address of the machine on which you ran the HTTP server.

```

> By default, doxygen also uses the `graphviz dot` tool to build some diagrams of the class, namespace, and module relationships. If the `dot` tool cannot be found then the output will be generated without diagrams. The doxygen installation page does not include instructions for downloading and installing `graphviz dot`.

Reviewer 2: ```suggestion
By default, doxygen uses the `graphviz dot` tool to build diagrams of the class, namespace, and module relationships. If the `dot` tool cannot be found then doxygen generates output without diagrams. The doxygen installation page does not include instructions for downloading and installing `graphviz dot`.
```

The doxygen output is intended for building documentation only for the public APIs and classes.

> For example, the output should not include documentation for `detail` or `/src` files.

Reviewer 2: ```suggestion
For example, the output should not include documentation for `detail` or `/src` files, and these directories are excluded in the `Doxyfile` configuration.
```
